File: go/jzoffer/verifyPostOrder_33.go
package jzoffer

import "math"

func verifyPostorder(postorder []int) bool {
	return recursionVeryPostOrder(postorder, math.MinInt32, math.MaxInt32)
}

func recursionVeryPostOrder(postOrder []int, min, max int) bool {
	if len(postOrder) == 0 {
		return true
	}
	root := postOrder[len(postOrder)-1]
	// 检查数字是否满足范围
	for _, num := range postOrder {
		if num < min || num > max {
			return false
		}
	}
	// 找到分隔点
	i := 0
	for i < len(postOrder)-1 && root > postOrder[i] {
		i++
	}

	return recursionVeryPostOrder(postOrder[:i], min, root) && recursionVeryPostOrder(postOrder[i:len(postOrder)-1], root, max)
}

File: java/jzoffer/VerifyPostorder_33.java
public class VerifyPostorder_33 {

    public boolean verifyPostorder(int[] postorder) {
        return recursion(postorder, 0, postorder.length - 1, Integer.MIN_VALUE, Integer.MAX_VALUE);
    }

    private boolean recursion(int[] postorder, int start, int end, int min, int max) {
        if (start >= end) return true;
        for (int i = start; i <= end; i++) {
            if (postorder[i] >= max || postorder[i] <= min) return false;
        }

        int root = postorder[end];
        int i = start;
        for (; i < end; i++) {
            if (postorder[i] > root) break;
        }

        return recursion(postorder, start, i - 1, min, root) && recursion(postorder, i, end - 1, root, max);
    }
}
